File: Striver A2Z/Step 4 - Binary Search [1D, 2D arrays, Search Space]/Lec 2 - BS on Answers/koko-eating-bananas.cpp

// https://leetcode.com/problems/koko-eating-bananas/

class Solution {
public:

    bool canFinish(vector<int>&piles, int h, int eatingSpeed){
        int time = 0;
        for(auto pile : piles){
            time += ceil((double)pile/eatingSpeed);
        }
        return (time <= h) ? true : false;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        int low = 1;
        int high = *max_element(piles.begin(), piles.end());

        while(low < high){
            int mid = (low + high)>>1;
            if(canFinish(piles, h, mid)) high = mid;
            else low = mid+1;
        }

        return low;
    }
};

File: Neetcode-150-js/Binary Search/koko-eating-bananas.js

// https://leetcode.com/problems/koko-eating-bananas/

/**
 * @param {number[]} piles
 * @param {number} h
 * @return {number}
 */

var canFinish = function(piles, h, eatingSpeed){
    let time = 0;
    for(let i = 0; i < piles.length; i++){
        time += Math.ceil(piles[i]/eatingSpeed);
    }

    console.log("time", time);
    return (time <= h) ? true : false;
}

var minEatingSpeed = function(piles, h) {
    
    // let start = Math.min(...piles);
    let start = 1;
    let end = Math.max(...piles);

    while(start < end){
        let mid = Math.floor((start+end)/2);
        console.log(start, end, mid);
        if(canFinish(piles, h, mid)) end = mid;
        else start = mid+1;
    }

    return start;
};